libraries/AR_Motors/AP_MotorsUGV.cpp

```diff
@@ -213,8 +220,101 @@ void AP_MotorsUGV::set_throttle(float throttle)
+    ** (throttle_chk < 0)  backward mode anyway
+    */
+    if ((throttle_chk < 0.0f) || ((throttle_chk <= 0.0f) && (_esc_backward_unlock_active) ))
+    {
+        const uint32_t now = AP_HAL::millis();
+        
+        switch (_esc_backward_unlock_state)
+        {
+            case UnlockState::UNLOCK_WAITING:
+                _esc_backward_unlock_simulated_signal=0;
+                _throttle = 0.0f;
+                _last_set_throttle_call = now;
+                _esc_backward_unlock_active = true;
+                _esc_backward_unlock_state = UnlockState::UNLOCK_STEP_DEC_TO_MAX_NEG;
+            break;
+
+            case UnlockState::UNLOCK_STEP_DEC_TO_MAX_NEG:
+                if ((now - _last_set_throttle_call) >= 20)
+                {
+                    _esc_backward_unlock_simulated_signal-=2.0f;
+                    _throttle = _esc_backward_unlock_simulated_signal;
+                }
+                _last_set_throttle_call = now;
+
+                if (_esc_backward_unlock_simulated_signal <= - 98.0f)
+                {
+                    _esc_backward_unlock_simulated_signal = -100.0f;
+                    _esc_backward_unlock_state = UnlockState::UNLOCK_STEP_STAY_MAX;
+                }
+            break;
+            
+            case UnlockState::UNLOCK_STEP_STAY_MAX:
+                _throttle = -100.0f;
+
+                if ((now - _last_set_throttle_call) > 300) 
+                {
+                    _esc_backward_unlock_state = UnlockState::UNLOCK_STEP_INC_TO_ZERO;
+                    _last_set_throttle_call = now;
+                }
+            break;
+            
+            case UnlockState::UNLOCK_STEP_INC_TO_ZERO:
+                if ((now - _last_set_throttle_call) >= 20)
+                {
+                    _esc_backward_unlock_simulated_signal+=2.0f;
+                    _throttle = _esc_backward_unlock_simulated_signal;
+                }
+                _last_set_throttle_call = now;
+
+                if (_esc_backward_unlock_simulated_signal >= -2.0f)
+                {
+                    _esc_backward_unlock_simulated_signal = 0.0f;
+                    _esc_backward_unlock_state = UnlockState::UNLOCK_STEP_ZEROS_2;
+                    _last_set_throttle_call = now;
+                }
+            break;
+
+
+            case UnlockState::UNLOCK_STEP_ZEROS_2:
+                _throttle = 0.0f;
+
+                if ((now - _last_set_throttle_call) > 50)
+                {
+                    _esc_backward_unlock_state = UnlockState::UNLOCK_FINISHED;
+                    _last_set_throttle_call = now;
+                    _esc_backward_unlock_active = false;
+                }
+            break;
+
+            case UnlockState::UNLOCK_FINISHED:
+                _throttle = throttle_chk;
+            break;
+        }
+    }
+    else 
+    {
+        _throttle = throttle_chk;
+
+        if ((_esc_backward_unlock_state!=UnlockState::UNLOCK_WAITING)  && (_throttle > 0))
```

Comment:
do you need this if?

Reply:
the function of this "if condition" is the following scenario:
1- press throttle backward i.e. < 0 then the state machine starts and the code starts to send the unlock sequence.
2- release throttle i.e. throttle = zero. then nothing should happen.
3- press throttle backward again. the unlock sequence should not restart again. _throttle should be equal to throttle value.
The condition prevents resetting state to "UnlockState::UNLOCK_WAITING" hence the sequence above occurs.

If we remove this condition then at state 2 then _esc_backward_unlock_state=UnlockState::UNLOCK_WAITING and when I relieve throttle state machine will reset.

Reply:
My read was that we can only get into this else when`throttle_chk` is larger than 0. On 309 your setting `_throttle = throttle_chk` so now `_throttle` must also be larger than 0. The check of `_esc_backward_unlock_state` does not actually do anything except prevent repeated setting. So I thought you might be able to remove the if and just leave `_esc_backward_unlock_state = UnlockState::UNLOCK_WAITING;`

Possibly I have missed something, GitHub is not great for such evaluations.

Reply:
any comment on this?

Reply:
@HefnySco

Reply:
The logic is 
1- driver pressed back to unlock so the unlock mechanism starts 
2- driver released the throttle
3- driver pressed back again .... 
then no need to initialize the state machine again of esc releasing as long as the driver didnt press forward throttle.
so keep state 
_esc_backward_unlock_state =UnlockState::UNLOCK_FINISHED;
rather than making it 
_esc_backward_unlock_state =UnlockState::UNLOCK_WAITING;